### PR TITLE
OOP school library: preserve data

### PR DESCRIPTION
**Pull request by NasKhalil**
this branch includes the following features implemented:

- On startup of your app load all data from files. Remember to handle the case when one or more of the files are missing.
- On exit from your app save all the data (books, people, and rentals) into JSON files. One per type of data you are storing and named accordingly (e.g. `rentals.json`)

### CODE REVIEW

File: book.rb [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
+require 'securerandom'
 
 class Book
   attr_accessor :title, :author
-  attr_reader :rentals
+  attr_reader :id, :rentals
 
-  def initialize(title, author)
+  def initialize(title, author, id: nil)
+    @id = id.nil? ? SecureRandom.uuid : id
```
Comment (dasileker) [inline book.rb:9]: - [ ] Please use Random.rand(1..1000) to get a proper id number for the student and the teacher.

**Review by dasileker — CHANGES_REQUESTED** [review]
### **Hi @NasKhalil,** 

### **Good job so far!**
There are some issues that you still need to work on to go to the next project but you are almost there!

## **Highlights**

- Everything is good 👌


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: book.rb (hunk 1 above)
Comment (ahangarha) [reply book.rb:9]: You mean a random number between 1 to 1000 is more proper than uuid? @dasileker

Comment (NasKhalil) [reply book.rb:9]: @dasileker Hi! Is using a random number between 1 and 1000 or uuid one of the project requirements ??

Comment (dasileker) [reply book.rb:9]: sorry the notification was off

Comment (dasileker) [reply book.rb:9]: hello, @NasKhalil @ahangarha right now you are securing the ID with numbers and letters, but the id should right now be a simple integer that can the user select to attribute the rental.

Comment (dasileker) [reply book.rb:9]: it can be done without the uuid

Comment (ahangarha) [reply book.rb:9]: You mean we introduce bug to the software? We have done something more secure.
I think this is blocking our PR for a reason which is not related to the project requirement. It is also asking us to make the code worse than better.

We can still wait for a good reason to change the code.

Comment (dasileker) [reply book.rb:9]: > You mean a random number between 1 to 1000 is more proper than uuid? @dasileker

for now yes

Comment (ahangarha) [reply book.rb:9]: I just request to share a good reason for this. Why I should leave a de-facto standard way of making unique id and use a random number between 1 to 1000? Just to make it simpler?

Please share a good reason for making the app more buggy, we will do it.

We have not violated any requirements. If needed, there are cases to ask for apply DRY principle. But this one to me doesn't look a valid change request.

**Review by Davidosky007 — COMMENTED** [review]
Hello @NasKhalil @ahangarha  👋,

_This is an additional review as a result of the second opinion request._

- Kindly note that it's okay to use any means to generate a unique ID as long as it doesn't break the app functionality, Apps with a larger code base ( Facebook and co) using the  Random.rand(1..1000) to generate unique ID might break as its likely similar ID is generated, but for apps with small code base like this is ok 👍 
- This request isn't part of the project requirement and as a result, should have been OPTIONAL ( Unless you found out that the method used to generate ID breaks the app) 

As this was the only reason that your project was not approved it will be approved as soon as possible.
Please ask for another review in your Dashboard and the next Code Reviewer will mark it as approved right away :)

Keep rocking 🚀!

**Comment by ahangarha** [comment]
Thanks @Davidosky007 

Also I want to say thanks to @dasileker for review. I know review gets tricky at some point. I know he was trying to do his bests.
Have a good day :tulip:

**Review by Mihndim2020 — APPROVED** [review]
# STATUS: APPROVED!!!  🎉 🎉 🎉 :champagne: :champagne: :champagne:
Hi Team, 

Your project is complete! Great work on making the changes requested by the previous reviewer. There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 :champagne: :champagne: :champagne:

Congratulations-Memes-FEATURED

## Optional suggestions

_Check the comments under the review._

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by ahangarha** [comment]
Thanks @Mihndim2020 
Just as you can check, this like is not accurate:

> Great work on making the changes requested by the previous reviewer.

And we know why ;)

Have a good day :white_flower:

**Pull request merged**